src/main/java/org/elasticsearch/plugins/repository/URLPluginRepository.java

```diff
@@ -0,0 +1,208 @@
+import java.util.List;
+
+import static org.elasticsearch.plugins.repository.PluginDescriptor.Builder.pluginDescriptor;
+
+public class URLPluginRepository implements PluginRepository {
+
+    private final String name;
+    private final Collection<String> urls;
+    private final PluginDownloader downloader;
+
+    protected URLPluginRepository(String name, Collection<String> urls, PluginDownloader downloader) {
+        this.name = name;
+        this.urls = urls;
+        this.downloader = downloader;
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public Collection<PluginDescriptor> find(String name) {
+        ImmutableList.Builder<PluginDescriptor> plugins = ImmutableList.builder();
+
+        // It could be smarter than that (like a real search by name) but for now
+        // it resolves all URLs and downloads every plugin if found.
+        if (urls != null) {
+            for (String pattern : urls) {
+                try {
+                    PluginDescriptor plugin = download(pattern, null, name, null);
+                    if (plugin != null) {
+                        plugins.add(plugin);
+                    }
+                } catch (IOException e) {
+                    // The find method ignore the download exception,
+                    // it should be reported by the listener.
+                }
+            }
+         }
+        return plugins.build();
+    }
+
+    @Override
+    public PluginDescriptor find(String organisation, String name, String version) {
+        if (urls != null) {
+            for (String pattern : urls) {
+                try {
+                    PluginDescriptor plugin = download(pattern, organisation, name, version);
+                    if (plugin != null) {
+                        return plugin;
+                    }
+                } catch (IOException e) {
+                    // The find method ignore the download exception,
+                    // it should be reported by the listener.
+                }
+            }
+        }
+        return null;
+    }
+
+    private PluginDescriptor download(String pattern, String organisation, String name, String version) throws IOException {
+        if (downloader == null) {
+            throw new PluginRepositoryException(name(), "No downloader defined for this plugin repository");
+        }
+
+        String url = convert(pattern, organisation, name, version);
+        if (url != null) {
+            Path downloaded = downloader.download(url);
+            if (downloaded != null) {
+                PluginDescriptor plugin = pluginDescriptor(name).loadFromZip(downloaded).artifact(downloaded).build();
+
+                // Check that the downloaded plugin corresponds to what has been asked for
+                if (StringUtils.equals(name, plugin.name())) {
+                    return plugin;
+                } else {
+                    Files.deleteIfExists(downloaded);
```

Comment:
Does it mean that we can ask for a given file but download something else?
Looks weird to me. If this could happen, should we log that information?